include/deal.II/physics/notation.h

```diff
@@ -0,0 +1,1550 @@
+              return std::make_pair(2u,0u);
+            case (8):
+              Assert(symmetric == false,
+                     ExcIndexRange(component_n,0,SymmetricTensor<2,3>::n_independent_components));
+              return std::make_pair(2u,1u);
+            default:
+              break;
+            }
+
+          AssertThrow(false, ExcInternalError());
+          return std::pair<unsigned int, unsigned int> ();
+        }
+
+
+        /**
+         * Return the scaling factor to be applied to the
+         * entry in the condensed vector.
+         */
+        template<int dim>
+        double
+        vector_component_factor (const unsigned int component_i,
+                                 const bool         symmetric)
+        {
+          if (symmetric == false)
+            return 1.0;
+
+          if (component_i < dim)
+            return 1.0;
+          else
+            return numbers::SQRT2;
+        }
+
+
+        /**
+         * Return the scaling factor to be applied to the
+         * entry in the condensed matrix.
+         */
+        template<int dim>
+        double
+        matrix_component_factor (const unsigned int component_i,
+                                 const unsigned int component_j,
+                                 const bool         symmetric)
+        {
+          if (symmetric == false)
+            return 1.0;
+
+          // This case check returns equivalent of this result:
+          // internal::vector_component_factor<dim>(component_i,symmetric)*internal::vector_component_factor<dim>(component_j,symmetric);
+          if (component_i < dim && component_j < dim)
+            return 1.0;
+          else if (component_i >= dim && component_j >= dim)
+            return 2.0;
+          else // ((component_i >= dim && component_j < dim) || (component_i < dim && component_j >= dim))
+            return numbers::SQRT2;
+        }
+
+      }
+
+
+      template<int,typename Number>
+      Vector<Number>
+      unroll_to_vector (const Number &s)
+      {
+        Vector<Number> out (1);
+        const unsigned int n_rows = out.size();
+        for (unsigned int r=0; r<n_rows; ++r)
+          out(r) = s;
+        return out;
+      }
+
+
+      template<int dim, typename Number>
+      Vector<Number>
+      unroll_to_vector (const Tensor<0,dim,Number> &s)
+      {
+        return unroll_to_vector(s.operator const Number &());
+      }
+
+
+      template<int dim, typename Number>
+      Vector<Number>
+      unroll_to_vector (const Tensor<1,dim,Number> &v)
+      {
+        Vector<Number> out (v.n_independent_components);
+        const unsigned int n_rows = out.size();
+        for (unsigned int r=0; r<n_rows; ++r)
+          {
+            const std::pair<unsigned int,unsigned int> indices = internal::indices_from_component<dim>(r,false);
+            Assert(indices.first < dim, ExcInternalError());
+            const unsigned int &i = indices.first;
```

Comment:
no need for a reference -- let the compiler stick that variable into a register, rather than having to keep pointers around.

Reply:
Does it make a difference since the indices are crated literally right above these aliases? I only wanted a sensible name for the variables instead of using `indices.first/second`.

Reply:
Oh, I'm fine with keeping the variable, but don't make it a reference:
```
  const unsigned int i = indices.first.
```
It's much easier for a compiler to deduce the flow of information without pointers and references.